File: Leetcode/0002-AddTwoNumbers/cpp/main.cpp
#include <iostream>

using namespace std;


struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};


class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode head;
        ListNode *node = &head;
        int carry = 0;
        while (l1 || l2) {
            int x = (l1 != NULL) ? l1->val : 0;
            int y = (l2 != NULL) ? l2->val : 0;
            int sum = x + y + carry;
            node->next = new ListNode(sum % 10);
            node = node->next;
            carry = sum / 10;
            if (l1)
                l1 = l1->next;
            if (l2)
                l2 = l2->next;
        }
        if (carry) {
            node->next = new ListNode(1);
        }
        return head.next;
    }
};


int main()
{
    Solution s;
    return 0;
}

File: Leetcode/0002-AddTwoNumbers/go/main.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	// Return immediately if one list is empty
	if l1 == nil {
		return l2
	} else if l2 == nil {
		return l1
	}
	// Dummy node to simplify the code
	dummy := &ListNode{}
	node := dummy
	carry := 0
	for l1 != nil || l2 != nil || carry != 0 {
		val1, val2 := 0, 0
		if l1 != nil {
			val1 = l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			val2 = l2.Val
			l2 = l2.Next
		}
		val := val1 + val2 + carry
		node.Next = &ListNode{
			Val: val % 10,
		}
		node = node.Next
		carry = val / 10
	}
	return dummy.Next

}

func main() {
}
